import subprocess
import os
from langchain.tools import tool

WORKSPACE_DIR = "./workspace"
SEED_REPO_DIR = "./bio_seeds"

@tool
def simulate_rna_seq_reads_art(
    reference_transcriptome_fasta: str, 
    num_reads_per_transcript: int = 50,
    read_length: int = 100
) -> tuple[str, str]:
    print(f"üß¨ Simulating RNA-Seq reads with ART...")
    output_prefix = os.path.join(WORKSPACE_DIR, "art_sim_rna")
    if not os.path.exists(output_prefix):
        os.makedirs(output_prefix)
    r1_path = output_prefix + "1.fq"
    r2_path = output_prefix + "2.fq"
    cmd = [
        "conda", "run", "-n", "bio_agent_env", "art_illumina",
        "-ss", "HS25",  
        "-i", reference_transcriptome_fasta,
        "-p",  
        "-l", str(read_length),
        "-f", str(num_reads_per_transcript),
        "-m", "300", 
        "-s", "50",  
        "-o", output_prefix,
    ]
    print(f"ART command: {' '.join(cmd)}")
    
    result = subprocess.run(cmd, capture_output=True, text=True)
    if result.returncode != 0:
        raise RuntimeError(f"ART Illumina failed: {result.stderr}\n{result.stdout}")
        
    print(f"‚úÖ RNA-Seq reads generated by ART at: {r1_path}, {r2_path}")
    return r1_path, r2_path

@tool
def simulate_rna_seq_reads_rsem(
    reference_transcriptome_fasta: str,
    total_reads: int = 1000000
) -> tuple[str, str]:
    print("üß¨ Simulating RNA-Seq reads with RSEM...")
    ref_name = "rsem_ref"
    ref_path = os.path.join(WORKSPACE_DIR, ref_name)
    if not os.path.exists(ref_path + ".grp"):
        print(f"‚è≥ Preparing RSEM reference for {reference_transcriptome_fasta}...")
        cmd_prep = ["conda", "run", "-n", "bio_agent_env","rsem-prepare-reference", "--quiet", reference_transcriptome_fasta, ref_path]
        subprocess.run(cmd_prep, check=True)
    output_prefix = os.path.join(WORKSPACE_DIR, "rsem_sim")
    if not os.path.exists(output_prefix):
        os.makedirs(output_prefix)
    r1_path = output_prefix + "_1.fq"
    r2_path = output_prefix + "_2.fq"
    isoform_results_path = os.path.join(WORKSPACE_DIR, "dummy.isoforms.results")
    with open(isoform_results_path, 'w') as f:
        f.write("transcript_id\tgene_id\tlength\teffective_length\tFPKM\tTPM\tIsoPct\n")
    
    cmd_sim = [
        "conda", "run", "-n", "bio_agent_env",
        "rsem-simulate-reads",
        ref_path,
        isoform_results_path, 
        "0.1", 
        str(total_reads),
        output_prefix
    ]
    
    result = subprocess.run(cmd_sim, capture_output=True, text=True)
    if "Can not open" in result.stderr or result.returncode != 0:
        raise RuntimeError(f"RSEM simulation failed: {result.stderr}")

    print(f"‚úÖ RNA-Seq reads generated by RSEM at: {r1_path}, {r2_path}")
    return r1_path, r2_path
